rfc/indexing-design.md

```diff
@@ -0,0 +1,228 @@
+
+The queryable attributes are as follows.
+- Site: the location(s) where all blob data is available
+- Shard: the location(s) where a specific shard of data is available
+- Slice: the location(s) within a shard where a block of data is available
+
+Query results can be filtered to select only results that the user has access to, or for data supplied by specific data providers.
+
+### Query Specification
+
+```
+match = {
+  subject=[myID, myOrgID],
+  location=spBaseURL,
+}
+results = Query(multihash="bafy...", match)
+```
+
+### Query Operation Diagram
+
+![w3up-query-diagram](w3up-query-diagram.png)
+
+## Indexing Components
+
+The indexing components are intended for deployment on the w3up gateway. They may also be deployed to the w3up client, where they will not be shared with other clients.
+
+### W3up Index
+
+The w3up index layer is responsible for reading a query specification, and then
+1. Creating a set of IPNI queries needed to get the requested index content
+2. Sending the queries to the w3up IPNI cache
+3. Filtering returned results
+4. Presenting results to the query client.
+
+### W3up IPNI Cache
+
+The w3up IPNI Cache is a cache that holds IPNI query results. The cache is:
+- Limited-size LRU. Discards the least recently used data when at storage capacity.
+- Temporary. Evicts items that have been cached for longer than the cache expiration time, regardless of last access. This allows changes to query results to be seen.
+- Negative (empty) result-aware. Caching empty results prevents more expensive IPNI queries for data that is not indexed.
+
+This is a read-through cache, meaning if it does not hold the request query results, then it forwards the query on to IPNI, and caches the results. This includes caching an empty response.
+
+There are two types of queries send to the cache. Each takes a data multihash, but returns a different result:
+1. shard/slice query. Returns one or more: shard CID, slice CID, offset, length (in shard).
+2. location query. Returns: location commitment of each shard. 
+
+By using both these query results, a specific portion of the blob data can be retrieved. The shard(s) are read from the locations in the location commitments, and data ranges within each shard are requested (when not reading an entire shard) by range requests using the offsets and lengths from the shard/slice data.
+
+If either shard or location results are not cached, then a query for the multihash is passed to the IPNI client.
+
+#### Expiration Times
+The shard/slice data does not change over the lifetime of the stored data, so this data can have a much longer expiration time. It could also have no expiration time and be explicitly removed from cache when the data is no longer stored, but this means that there need to be a communication path from data removal to all cache locations. So, better to just have an expiration time, making cache more independent.
```

Comment:
I'm not sure we need to evict information about data even if we don't store that data. It could be that location of the data is not in our system (e.g. shared privately) while index of it is shared publicly. That is to say I don't think we need to worry about removing shard/slice info and instead let the user perform manual eviction when they want to do so. Specifically I imagine user would delete CAR holding an index which could trigger eviction

Reply:
curious about a side issue here -- what do we do with two people who upload the same blob but publish different indexes? Seems like this would be relevant. Probably addressed if we add the "account paying" query param that the egress billing design calls for -- does make me wonder if we need to filter CID results not only on "are they W3UP" but also are "are they for the right person for this query?"